File: src/main/java/com/library/factories/UserFactory.java
package com.library.factories;

import com.library.users.Utilisateur;
import com.library.users.Etudiant;
import com.library.users.Enseignant;

/**
 * Factory Pattern - UserFactory
 * Crée les instances d'utilisateurs en fonction du type
 */
public class UserFactory {

    public enum UserType {
        ETUDIANT,
        ENSEIGNANT
    }

    /**
     * Crée un utilisateur du type spécifié
     * @param type Le type d'utilisateur à créer
     * @param id L'identifiant de l'utilisateur
     * @param nom Le nom de l'utilisateur
     * @param email L'email de l'utilisateur
     * @param additionalInfo Information supplémentaire (numéro étudiant ou département)
     * @return L'utilisateur créé
     */
    public static Utilisateur createUser(UserType type, String id, String nom, 
                                         String email, String additionalInfo) {
        switch (type) {
            case ETUDIANT:
                return new Etudiant(id, nom, email, additionalInfo);
            case ENSEIGNANT:
                return new Enseignant(id, nom, email, additionalInfo);
            default:
                throw new IllegalArgumentException("Type d'utilisateur inconnu: " + type);
        }
    }
}

File: js/src/factories/UserFactory.js
const Etudiant = require('../users/Etudiant');
const Enseignant = require('../users/Enseignant');

/**
 * Factory Pattern - UserFactory
 * Crée les instances d'utilisateurs en fonction du type
 */
class UserFactory {
    static UserType = {
        ETUDIANT: 'ETUDIANT',
        ENSEIGNANT: 'ENSEIGNANT'
    };

    /**
     * Crée un utilisateur du type spécifié
     * @param {string} type - Le type d'utilisateur (ETUDIANT ou ENSEIGNANT)
     * @param {string} id - L'identifiant de l'utilisateur
     * @param {string} nom - Le nom de l'utilisateur
     * @param {string} email - L'email de l'utilisateur
     * @param {string} additionalInfo - Information supplémentaire (numéro étudiant ou département)
     * @returns {Utilisateur} L'utilisateur créé
     */
    static createUser(type, id, nom, email, additionalInfo) {
        switch (type) {
            case UserFactory.UserType.ETUDIANT:
                return new Etudiant(id, nom, email, additionalInfo);
            case UserFactory.UserType.ENSEIGNANT:
                return new Enseignant(id, nom, email, additionalInfo);
            default:
                throw new Error(`Type d'utilisateur inconnu: ${type}`);
        }
    }
}

module.exports = UserFactory;
